### DISCUSSION TITLE
getting_started/first_2d_game/04.creating_the_enemy

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#82] [User-contributed notes]
# getting_started/first_2d_game/04.creating_the_enemy

Now it's time to make the enemies our player will have to dodge. Their behavior will not be very complex: mobs will spawn randomly at the edges of the screen, choose a random direction, and move in...

https://docs.godotengine.org/en/stable/getting_started/first_2d_game/04.creating_the_enemy.html

**Comment by ThomasCrane** [comment]
C# users *must* change the connections under the Node tab to C# nomenclature, not the default it provides you.

Change _on_visible_on_screen_notifier_2d_screen_exited to OnVisibleOnScreenNotifier2DScreenExited, or, if you've already created the script as guided, you can select the Pick button to the right of the text field when editing the connection, and then select the appropriate method.

**Reply by Xiarno** [reply]
I'm unable to pick as instructed or don't understand what you mean.

**Comment by Wadahell** [comment]
I need help with the enemy animations.
They appear when i'm fitting the hitbox, but not when I run the scene

**Reply by BastardSeagull** [reply]
There's a pause/stop button in the Animation panel. If you don't click it, the animation will keep playing while you're in the 2D scene view.

**Reply by Wadahell** [reply]
That's not the problem, it works on the 2d scene and dont on the game. 
And also, it doesnt randomize, it only uses the last animation i edit, like if i click on Swim, he only shows the swimming mob on the game and without the animation
Also got a weird Break point in the line 
$HUD.show_message("Get Ready")

my game doesnt run if this is in the code

**Reply by SeanSan06** [reply]
This is late but it could be because your running the player scene and not the mob scene

**Reply by oulfik** [reply]
I'm also late to the party: I had the exact same problem. In my case I forgot to attach the mob script to the node. You have to check this in the Inspector view at the bottom.

**Reply by IrwinKwan** [reply]
I had the same problem as well. I fixed it by enabling "Autoplay on load", which looks like an arrow with an "A" in it.

**Reply by AetherialAvenger** [reply]
Insert $AnimatedSprite2D.play() below $AnimatedSprite2D.animation = mob_types.pick_random()

$AnimatedSprite2D.animation = mob_types.pick_random() only tells the node to SET a specific animation

You still have to tell it to PLAY the animation with .play()

**Comment by PamProg** [comment]
When you say to "To align the shape with the image", not all sprites have the same size, but we only use 1 AnimatedSprite2D, shouldn't we use one for each ?

**Reply by BastardSeagull** [reply]
I just based it on the swimming one, but you probably could do a different one for each. I'm new to Godot, but I can't think of a way to do it without creating 3 separate mobs and updating on_mob_timer_timeout() to select one instead of having the selection in the mob script.

Might be a fun way to try to expand on the project!

**Reply by SolarLunix** [reply]
I created two 2D Collisions and called them Collision_WalkSwim and Collision_Fly I leave Collision_WalkSwim alone. Collision_Fly just goes around the legs.

```
func _ready() -> void:
	var mob_types = $AnimatedSprite2D.sprite_frames.get_animation_names()
	var mob_type = mob_types[randi() % mob_types.size()]
	$AnimatedSprite2D.play(mob_type)
	if mob_type == "fly":
		$Collision_Fly.disabled = false
	else:
		$Collision_Fly.disabled = true
```

Hopefully it works. 

Edit: Appears to work just fine :D

**Reply by PrizmaticAtmos** [reply]
Thanks for the idea, here's another way in case it's helpful. I added a second collision map and named it CollisionFly.

```Gdscript
func _ready() -> void:
	var mob_types : Array
	mob_types = Array(animated_sprite_2d.sprite_frames.get_animation_names())
	animated_sprite_2d.animation = mob_types.pick_random()
	
	if animated_sprite_2d.animation == "fly":
		collision_fly.disabled = false
		collision_walk_swim.disabled = true
	else:
		collision_fly.disabled = true
		collision_walk_swim.disabled = false

	animated_sprite_2d.play()
```

**Reply by AfiqAiman24** [reply]
@PrizmaticAtmos Nice one, thanks man! Here, I found a shorter (maybe cleaner?) version based off of yours:
```GDScript
func _ready() -> void:
	var mob_types : Array
	mob_types = Array(animated_sprite_2d.sprite_frames.get_animation_names())
	animated_sprite_2d.animation = mob_types.pick_random()
	
	var is_fly_animation: bool = animated_sprite_2d.animation == "fly"
	collision_fly.disabled = not is_fly_animation
	collision_walk_swim.disabled = is_fly_animation

	animated_sprite_2d.play()
```

**Comment by MoroAstray** [comment]
What does the Collision Mask mean? It just says to uncheck the 1 but it feels really random with no explanation

**Reply by BastardSeagull** [reply]
The collision mask is where it looks for what it's colliding with. So by unchecking 1, it's not bothering to check if it's touching anything.

**Reply by vmorris** [reply]
What do the other mask values do?

**Reply by Calinou** [reply]
> What do the other mask values do?

By default, only the layer 1 is used. You can add objects to other layers to have more granular collision behavior (e.g. have the player collide with a certain object, but have enemies phase through this object instead).

Other mask values will only have a noticeable effect if you have at least one object on that layer.

**Reply by JoemanJ** [reply]
To expand a bit on this explanation, you can think of collision layers and mask this way:
If an object is active on LAYER 1, that means it can "hit" objects in layer 1;
If an object is active on MASK 1, that means it can "be hit by" objects in layer 1.

The way the collisions are configured in this tutorial basically means that "enemies can hit a player (because the enemies are active on LAYER 1 and the player is active on MASK 1) but an enemy can't hit another enemy (because the enemies aren't active in any MASK)"

**Reply by BytesBlade** [reply]
This explanation needs to be in the doc

**Reply by FrackinFamous** [reply]
If I am the Player object:

    Collision layer is my reality. It's where I live physically. 

    Mask layer is anything I want to interact with or have interact with me...walls, coins enemies. 

    (Any layers I don't have masked I phase through)

The explanation is in the docs, in great detail here:

https://docs.godotengine.org/en/stable/tutorials/physics/physics_introduction.html#collision-layers-and-masks

Everything in the tutorial is meant to spur us to research deeper into the docs.

**Comment by YoungBacC** [comment]
Make sure to set the collision shape 2D on the mob or else it will not work when it hits the player. I know this is obvious, but I forgot LOL.

**Comment by vmorris** [comment]
> Connect the screen_exited() signal of the VisibleOnScreenNotifier2D node to the Mob and add this code:

This has me going :hmm:

I know you just explained how to do this in the previous section, but maybe you could give alittle more context?

**Reply by vmorris** [reply]
Like... what is "queue_free"?

**Reply by vmorris** [reply]
Answering my own question here.. maybe this was given in a previous section of the getting started?

> Unlike with Object.free(), the node is not deleted instantly, and it can still be accessed before deletion. It is also safe to call queue_free() multiple times. Use Object.is_queued_for_deletion() to check if the node will be deleted at the end of the frame.

>  Note: The node will only be freed after all other deferred calls are finished. Using this method is not always the same as calling Object.free() through Object.call_deferred().

**Reply by Andrewjdesigner** [reply]
This part confused me too. For people like me, taking breaks between sections, `screen_exited()` is a signal built into the `VisibleOnScreenNotifier2D` node. Select `VisibleOnScreenNotifier2D` in the scene tree. Then, click the Node tab next to the Inspector. There you'll find the `screen_exited()` signal. Right-click and select connect to connect it to your `Mob` node.

**Reply by madopal** [reply]
As of March 2026 (v4.6.1):
1. The tab is now called `Signals` next to `Inspector`, not `Node`
2. If you don't attach the script to the base node (or you create the script in a different manner), the connect will fail even if you have the routine already created in your  script  (`_on_visible_on_screen_notifier_2d_screen_exited`).

**Comment by GalaxyBangBang** [comment]
Why do you use RigidBody2D as the root node of mob, but use Area2D as the root node of player? Are there any differences? Thanks for any responses.

**Reply by GalaxyBangBang** [reply]
I find a blog and it answers my problem, https://code.luasoftware.com/tutorials/godot/godot-area2d-vs-staticbody2d-ve-rigidbody2d-vs-kinematicbody2d. In short, you use force to control rigidbody but you use setPos to control area 2D.

**Reply by JoemanJ** [reply]
In general, you should choose the node type what will be "easier to control" for your objects.

In this case, because the way the enemies move (which is defined in the next section), it is easier to just "give them a push" in some direction and have them moving on their own (or rather, have the physics simulation move them for us) instead of update their positions ourselves like we did with the player, so we chose 'RigidBody2D' in their case.

You are correct in the idea that 'RigidBody2D' is to be "moved by aplying force" and that 'Area2D' is to be "positioned" (props to you for doing your own research and posting your answer to others btw). You can think of an 'RigidBody2D' object's movement as a ball being pushed around, while a 'Area2D' object's movements would be more like the same ball being grabbed and moved

**Comment by kjosh** [comment]
In these examples I often see repeated uses of ```$AnimatedSprite2D```. As far as I understand it this is syntactic sugar as an alias for ```get_node("AnimatedSprite2D")```. 

In real projects, is it worth it to assign the node to a variable and reuse that instead of repeatedly calling ```get_node``` in a hot code path, or is the call so cheap that I shouldn't usually bother and prefer the better readability?

**Reply by Calinou** [reply]
> In real projects, is it worth it to assign the node to a variable and reuse that instead of repeatedly calling get_node in a hot code path, or is the call so cheap that I shouldn't usually bother and prefer the better readability?

Yes, the `get_node()` call is cheap but it's not free. Also, assigning it to a variable makes it easier to benefit from [static typing](https://docs.godotengine.org/en/stable/tutorials/scripting/gdscript/static_typing.html), which can provide further performance benefits thanks to typed instructions:

```gdscript
# NOTE: The type must be mentioned explicitly.
# Type inference (`:=`) cannot be used for a precise type hint (unless casting with `as`).
# Otherwise, it will just be a Node type hint.
var sprite: AnimatedSprite2D = $AnimatedSprite2D
```

**Comment by madddmadd** [comment]
I set my gravity scale to 0 on the rigid body and the enemy still falls and i cant seem to figure out why

**Comment by jeroenriess** [comment]
The _on_body_entered function is never being called. I followed this tutorial closely. 
Here's information about my setup which I think is important for collisions:
- Player is type Area2D
- Player has CollisionShape2D drawn around the entire body
- Player has collision layer 1, mask 1
- Mob is of type RigidBody2D
- Mob also has a CollisionShape2D drawn around entire body
- Mob has collision layer 1,  and no mask

I set "Debug > Visible collision shapes" so that I can clearly see the shapes overlapping.
I have a print statement in the _on_body_entered, and nothing is printing, so this function is clearly not running. 
The _on_body_entered function seems to be linked correctly to the signal. 
I am even printing "body_entered.is_connected(_on_body_entered)" at every cycle and get true as output.

Any other ideas to debug this issue? Thanks.

**Reply by jeroenriess** [reply]
I found my issue... 
In the Player start() function, I wrote: $CollisionShape2D.disabled = true instead of false.

**Reply by Hamarius** [reply]
Thank you so much, I also struggled with this

**Comment by PaulRS747** [comment]
When I type extends Rigidbody2D in mob, if it already has extends Area2D then it generates an error, but if I change the extends Area2D to Rigidbody2D it makes the same change also in player.  I do not know how to proceed.

**Comment by kotsaris** [comment]
Keep in mind that the OnVisibleOnScreenNotifier2DScreenExited method has to be public for the Pick button to work on the Connect dialog. If you have it Private, like in the Tutorial above, it won't be detected.

**Comment by somebarcode** [comment]
For some reason the enemy Mobs collide and bounce off each other's Mobs.

How can the Mobs fly around without crashing next to the other ones? Not sure where I've caused them to collide.

**Reply by stefangorneanu97** [reply]
Make sure you Select "RigidBody2D" > Collision > Click on Mask (1) so it is NOT highlighted. This prevents them from colliding with each other.

**Comment by ElectricZax** [comment]
In CollisionObject2D we disable "Mask 1", so that the enemies can go through one another, correct?
The "Mask 1" is enabled on the Player, so that it gets hit...
How can I specify or differentiate from hits from Mask 1 or Mask 2, so I can create different actions for different situations?

For example something like _on_body_entered.mask_1() ?

**Comment by ElectricZax** [comment]
I tried to switch up things a bit, but have a problem...
Since the Mob scene (RigidBody2D) has "body_entered" as signal, I tried to setup the body entered signal in Mob instead of in Player. But it doesn't work. It looks like the "body_entered" is ignored and nothing happens. Any ideas why?

**Comment by FunkyToc** [comment]
Note that the "%" means "modulo operator" (and not a specific/weird keyword thing)
>randi() % mob_types.size() 

So, you can read this as:
- ```randi()``` take a random interger between 0 and 2^32
- ```%```  modulo (math operator, like * / + -)
- ```mob_types.size()``` size of our array

You could use this as an equivalent without using modulo:
```randi_range(0, mob_types.size() - 1)```

**Comment by RockyKev** [comment]
The demo project and the tutorial has different code and it should be synced.

In the tutorial:
```gdscript
func _ready() -> void:
	var mob_types = $AnimatedSprite2D.sprite_frames.get_animation_names()
	$AnimatedSprite2D.play(mob_types[randi() % mob_types.size()])
```

Where the [demo project's mob.gd](https://github.com/godotengine/godot-demo-projects/blob/master/2d/dodge_the_creeps/mob.gd) is:

```gdscript
func _ready():
	$AnimatedSprite2D.play()
	var mob_types = Array($AnimatedSprite2D.sprite_frames.get_animation_names())
	$AnimatedSprite2D.animation = mob_types.pick_random()

```

**Comment by Selenestica** [comment]
A .gif showing how the mobs should behave at the end of this part of the tutorial would be super helpful. When I run this script, I have one mob in the upper left corner of the screen doing it's animation. I'm not sure if this is correct or not.

**Reply by Crysteier** [reply]
Edit for I gave the worng anser before:
the mob will not behave like anything. In this module you just setup how the mob will behave in the next section of this tutorial. Just keep on going and you will see it.

**Comment by kjenney** [comment]
In addition, under the CollisionObject2D section just beneath the RigidBody2D section, expand the Collision group and uncheck the 1 inside the Mask property. This will ensure the mobs do not collide with each other.
This took me a little bit to figure out because I had already named RigidBody2D to Mob.

**Comment by jpkloess** [comment]
I'm running into a roadblock with this tutorial in Godot 4.3. I'm actually on the following step of the tutorial (The main game scene), and my game is crashing every time I run it with the error: "Invalid call. Nonexistent function 'get_animation_names' in base 'AnimatedSprite2D'."

The mobs do not spawn along my path or anywhere else, and the game just crashes. Is there a new getter I need to make this work?

**Reply by jpkloess** [reply]
Edit: I can confirm the issue is the animation. If I comment out the code in "func _ready()" the mobs spawn as intended. No idea why the animation get is not working though.

**Reply by boertv** [reply]
Did you forget sprite_frames maybe? Like this `$AnimatedSprite2D.sprite_frames.get_animation_names()`

**Reply by phicar2** [reply]
I had the same error...and my mistake was that I missed the 's' in 'names'
I had: $AnimatedSprite2D.sprite_frames.get_animation_name()
instead of: $AnimatedSprite2D.sprite_frames.get_animation_names()

**Comment by feihuan1** [comment]
```gdscript
extends RigidBody2D

func _ready() -> void:
	var mob_types = $AnimatedSprite2D.sprite_frames.get_animation_names()
	$AnimatedSprite2D.play(mob_types[randi() % mob_types.size()])

func _on_visible_on_screen_notifier_2d_screen_exited() -> void:
	$CollisionShape2D.set_deferred("disabled", true)
	queue_free()
```

is it neccesary use set_deffered here?

**Reply by Calinou** [reply]
Yes: https://docs.godotengine.org/en/latest/classes/class_collisionshape2d.html#class-collisionshape2d-property-disabled

(This is also the case in 3D.)

**Comment by TheLongCorgi** [comment]
For anyone else looking for why the animation isn't playing: You need to add 
```
$AnimatedSprite2D.play()
```
under 
```
func _ready():
```
It should look like:
```
func _ready():
	$AnimatedSprite2D.play()
	var mob_types = Array($AnimatedSprite2D.sprite_frames.get_animation_names()) #sets an array called "mob_types" with a list of items that are the names of the animations
	$AnimatedSprite2D.animation = mob_types.pick_random() #picks a random mod type from the array
```

**Reply by Calinou** [reply]
This was added to the documentation in https://github.com/godotengine/godot-docs/pull/10752, but it hasn't been cherry-picked to `stable` yet.

**Reply by atollk** [reply]
As an alternative, enable "Autoplay On Load" on the animation frames.

**Comment by atollk** [comment]
I feel like this guide is missing a step to resize the VisibleOnScreenNotifier2D node. When I created it, it was like a 10x10 rectangle which caused the the mobs to despawn at the edge of the screen while still visible. The node rect has to be resized to include the entire mob sprite.

**Comment by WIZN0DE** [comment]
Shouldn't we add $AnimatedSprite2D.play() here? Without this the mob is not animating in my Main scene.

func _ready():
	var mob_types = Array($AnimatedSprite2D.sprite_frames.get_animation_names())
	$AnimatedSprite2D.animation = mob_types.pick_random()
    $AnimatedSprite2D.play()

**Reply by Calinou** [reply]
This was already discussed above: https://github.com/godotengine/godot-docs-user-notes/discussions/82#discussioncomment-12506978

**Reply by WIZN0DE** [reply]
Yes sorry, I missed it.

**Comment by TheDeadMeasure** [comment]
I'm trying to set up my mob nodes, and to make the children unselectable (per "Don't forget to set the children so they can't be selected, like you did with the Player scene."), but I don't have that icon (or any icons) showing up beside the mob root node. I didn't have a problem with setting up the Player scene. I can't figure out how to get that icon to show up, or to find it in options somewhere.

**Reply by TheDeadMeasure** [reply]
I continued on, and when I moved the AnimatedSprite2D away from the other nodes (because I hadn't grouped them), then hit control z to bring the sprite back on top of the rest, the icon finally showed up in the window bar. I'm not sure if I did something wrong to begin with.

**Reply by suspiciousRaccoon** [reply]
This is under the 2D portion on the top of the screen. You were probably on the Script part.

**Comment by zancorin** [comment]
Hmm, the CollisionShape2D can be aligned nicely with the swim and walk sprite animations, but not with the fly one... I guess since this is an earlier tutorial, I shouldn't worry about accuracy too much?

**Comment by Nowell5** [comment]
So my mob is using the animation in the game, but the mob is just flying across the screen. What's up with that? The mob is not going through it's animation cycle properly and as soon as it enters the area it instantly falls and does not go across the screen as intended.

**Reply by Nowell5** [reply]
Alright so I fixed it, I went into project settings went to physics, clicked on 2D and set my default gravity to 0.0. It was set at 980 that is why my mobs were instantly dropping to the ground as soon as they spawn into the map !

**Comment by jonpepler** [comment]
Adding a new Mob scene in this page has really confused me. I can't tell what I'm doing wrong, but when I add a new scene, it appears to delete the Player scene and replace it as the root node? Or at least I can't see it in the Scene list anymore. I'll keep trying and will update this comment if I figure it out.

**Reply by jonpepler** [reply]
Aand I understand now. For anyone else struggling with this, new scenes exist in separate tabs at the top of the window. That wasn't obvious to me for whatever reason.

**Comment by nkocodes** [comment]
_"As in the Player scene, add a CapsuleShape2D for the collision. To align the shape with the image, you'll need to set the Rotation property to 90 (under "Transform" in the Inspector)."_

What does this part mean? The rotation property in CollisionShape2D? When I set the rotation to 90, it doesn't follow the animation.

**Reply by jonpepler** [reply]
I believe it means CapsuleShape2D, it also has a transform property.

**Comment by maruwapo-netizen** [comment]
Hey, love this tutorial but you really need to repeat stuff instead of asking us to go back to earlier parts of the tutorial. It's really frustrating to have to constantly be skipping around between sections, you can't expect people to remember everything, especially if its their first time using the software.

Example: "Don't forget to set the children so they can't be selected, like you did with the Player scene."

You could just reexplain how to make it so children can't be selected, that would help it stick.

**Reply by Calinou** [reply]
Pull request opened:

- https://github.com/godotengine/godot-docs/pull/11268

**Reply by FrackinFamous** [reply]
A little searching goes a long way towards ingestion of the information. Just following each perfectly curated and overexplained step slows one's learning down considerably. Additionally, easy first tutorials will make you think programming is fun. :)

**Reply by WargamingR** [reply]
Hey programming IS fun. But it's also frustrating, satisfying and a challenge. Sometimes all at the same time :)

**Comment by shallow-stars** [comment]
So 'AnimatedSprite2D.sprite_frames.get_animation_names()' yields an array of names.
But then in the example code, this Array is wrapped in Array(), without context for why?

**Comment by mbrinton91** [comment]
I'm gonna be honest, the quality dropped off a cliff with this chapter of the documentation. Things were being explained pretty well and the instructions were specific and thorough before this. Now with stuff like telling us to uncheck the "1" under mask with no explanation as to why / what that does exactly, telling us to set up AnimatedSprite2D the same way we did for the player (WDYM? Needs to be more specific) and telling us to change values without telling us where they are or why we're changing them... At this point I think I'm just going to stop walking through the docs, drop this whole project and find another tutorial that does a better job of walking through things.

**Reply by Calinou** [reply]
Please open an issue on https://github.com/godotengine/godot-docs/issues to request improvements to the documentation. Be as specific as possible in the issue text, so we know what to improve exactly.

**Comment by accompliceofthemouse** [comment]
It won't allow me to select CollisionObject2d. it says "This class can't be initiated"

**Reply by accompliceofthemouse** [reply]
sorry I mean instantiated

**Comment by vistuleB** [comment]
The LITTLE RED DOT CONTROL HANDLES on the CollisionShape2D are very small & hard to grab on MacOS.

Go to Godot -> Editor Settings -> "General" tag -> Interface -> Touchscreen -> set "Scale Gizmo Handles" to 2 or above

👍

**Reply by WargamingR** [reply]
Are you using a laptop? I have a large monitor with my Mac Studio and it was ok, so it might be a function of screen size etc.

**Comment by tomas-chavarria** [comment]
I'm having an issue where the following warning is shown when I create the signal call method [Please note that I am using C#]:
```
Missing connected method 'OnVisibleOnScreenNotifier2DScreenExited' for signal 'screen_exited' from node 'Mob/VisibleOnScreenNotifier2D' to node 'Mob'.
```

**Comment by Falco-Boehnke** [comment]
var mob_types = Array($AnimatedSprite2D.sprite_frames.get_animation_names())

On this codeline, why do we need to do Array(...) if the get_animation_names does already return an array?

**Comment by LBartemius** [comment]
Thanks for the tutorial! Please everyone, one tip, remember that when you add "VisibleOnScreenNotifier2D" you must also **increase the size of the <ins>pink rectangle in the 2D tab to cover the Mob</ins>**. 

In my case, the rectangle was smaller than the Mob, so they got stuck at the border of the screen when playing the final game.

**Comment by MKBSP** [comment]
it says:
"Click Scene -> New Scene from the top menu and add the following nodes:
- RigidBody2D (named Mob)
-- AnimatedSprite2D
-- CollisionShape2D
-- VisibleOnScreenNotifier2D"

And: 
"In addition, under the CollisionObject2D section just beneath the RigidBody2D section, expand the Collision group and uncheck the 1 inside the Mask property. This will ensure the mobs do not collide with each other."

In the newest update there is no: "CollisionObject2D" node, only "CollisionShape2D"

**Reply by noy88** [reply]
I made the same mistake. It is under the Inspector for RigidBody2d, or "mob" for this section.

**Reply by noy88** [reply]
Article could use a slight correction, since the verbiage is talking about a section under the main object.

**Reply by Calinou** [reply]
Pull request opened:

- https://github.com/godotengine/godot-docs/pull/11867

**Comment by scravy** [comment]
I reduced the speed of the mobs, so my four year old could play this.  Some of them stopped entirely.  There is a global linear damp setting in project settings -> physics which has a linear damp of 0.1 by default apparently.  Setting the "Damp mode" to "replace" instead of "combine" made the problem go away.

**Comment by admin** [comment]
*Comment originally posted by [MajesticPink](https://github.com/MajesticPink):*

> An addendum to the part about adding CollisionShape2D to the mob scene:
> 
> *"As in the Player scene, add a CapsuleShape2D for the collision."*
> 
> This gives 1 collision shape for all 3 mob variants. However I have noticed that the mobs have different sizes (in this example only the flying mob is smaller than the other two. 
> 
> In an actual distributable game this would create a confusion layer where the visual representation doesn't match up with the colliding shape of the enemy. 
> 
> To remedy this tiny issue, one can add a dynamic collision shape scaling match (or switch) case to modify the scale of the collision shape for the randomly picked mob animation sprite. 
> 
> When initially picking out the collision shape under CollisionShape2D node's Inspector view, leave the selected shape as is without altering its shape in the 2D view panel. Still in the Inspector Tab expand the Transform > sub-section under Node2D section and change the Scale values from there to match the collision shape to individual mobs. 
> 
> To do so, go to AnimatedSprite2D and select 1 mob variant from animation tab at a time and go back to CollisionShape2D. 
> 
> Adjust Scale property for each until it best fits the mobs shape and write down the x and y values for each of them.
> 
> Put the Scale x and y back to 1.0 in the inspector as we will modify its value in the script.
> 
> In the script, after adding the tutorial's code in the _ready() function, adjust it to look similar to this code:
> 
> ```py
> func _ready() -> void:
> 	var mob_types = Array($AnimatedSprite2D.sprite_frames.get_animation_names())
> 	var picked_mob = mob_types.pick_random() # We store the randomly picked sprite in a custom variable
> 	$AnimatedSprite2D.animation = picked_mob # Associate it to the animation
> 	# Below, we match our picked_mob variable to see if it's the flying mob or other ("_" stands for default/other cases in gdscript)
> 	match picked_mob: 
> 		"fly":
> 			$CollisionShape2D.scale = Vector2(3.4, 2.7) # In case the random mob is of flying type we adjust the collision shape to conform it's visual shape
> 		_:
> 			$CollisionShape2D.scale = Vector2(3.6, 3.3) # In case its the other two we do the same with different values
> 	$AnimatedSprite2D.play() # Lastly we play it
> ```
> 
> This works for each child node instantiated and is pretty low cost since we aren't using an if-elif-else operator.
> 
> It gives more control in terms of customizing of mobs in the same scene and also introduces match operator :D
> 
> Godot Reference for match operator: https://docs.godotengine.org/en/latest/tutorials/scripting/gdscript/gdscript_basics.html#match
> 
> P.S.: works the same for C# you just gotta use different syntax since GDscript match is C# switch case: https://www.w3schools.com/cs/cs_switch.php

**Comment by NobodyAnswered** [comment]
```gdscript
extends RigidBody2D


# Called when the node enters the scene tree for the first time.
func _ready():
	var mob_types = Array($AnimatedSprite2D.sprite_frames.get_animation_names())
	$AnimatedSprite2D.animation = mob_types.pick_random()
	$AnimatedSprite2D.play()
	pass # Replace with function body.

func _on_visible_on_screen_notifier_2d_screen_exited():
	queue_free()
	
# Called every frame. 'delta' is the elapsed time since the previous frame.
func _process(delta: float) -> void:
	
	pass
```

This is the code, but it doesn't work, why? 

Also, this is literally the most unexplained, unclear tutorial  I have ever seen.
For example, what is this?:

"The last piece is to make the mobs delete themselves when they leave the screen. Connect the screen_exited() signal of the VisibleOnScreenNotifier2D node to the Mob and add this code:"

This is the first time signals are mentioned in the tutorial, and you expect us to know what to do without any pictures for reference and such? I think I will stick with Unity, at least it has better, more clear explanations, docs and tutorials.
